File: python/src/molplot/annotations.py
"""Chart annotations — portable Vega-Lite layers (mirrors core/src/annotations.ts).

Kinds:
  * ``scaleBar`` — a ``|———|`` bar in data coordinates
  * ``arrow``    — directed arrow (rule + tip) with optional label

Use :func:`with_annotations` on any unit/layered VL dict, or put an
``annotations`` list on a fence / RawChart payload (stripped before render).
"""

from __future__ import annotations

import math
from typing import Any, Literal, Mapping, Sequence, TypedDict

__all__ = [
    "ScaleBarAnnotation",
    "ArrowAnnotation",
    "Annotation",
    "annotation_layers",
    "with_annotations",
    "take_annotations",
]


class ScaleBarAnnotation(TypedDict, total=False):
    kind: Literal["scaleBar"]
    x: float
    y: float
    length: float
    x2: float
    y2: float
    label: str
    color: str
    strokeWidth: float
    orientation: Literal["horizontal", "vertical"]
    tick: float
    tickRatio: float


class ArrowAnnotation(TypedDict, total=False):
    kind: Literal["arrow"]
    x: float
    y: float
    x2: float
    y2: float
    label: str
    color: str
    strokeWidth: float
    tipSize: float


Annotation = ScaleBarAnnotation | ArrowAnnotation


_SERIF = "Times New Roman, Times, STIX Two Text, STIXGeneral, serif"


def _log_perp_cap(
    x: float, y: float, x0: float, y0: float, x1: float, y1: float, s: float
) -> dict[str, float]:
    import math

    dx = math.log(x1 / x0)
    dy = math.log(y1 / y0)
    n = math.hypot(dx, dy) or 1.0
    px, py = (-dy / n) * s, (dx / n) * s
    return {
        "x": math.exp(math.log(x) + px),
        "y": math.exp(math.log(y) + py),
        "x2": math.exp(math.log(x) - px),
        "y2": math.exp(math.log(y) - py),
    }


def _scale_bar_layers(a: Mapping[str, Any]) -> list[dict[str, Any]]:
    import math

    color = a.get("color") or "#14271d"
    sw = a.get("strokeWidth") or 1.8
    orient = a.get("orientation") or "horizontal"
    x2, y2 = a.get("x2"), a.get("y2")
    along = orient == "along" or (
        x2 is not None and y2 is not None and orient not in ("horizontal", "vertical")
    )
    length = float(a["length"]) if a.get("length") is not None else 0.0
    tick = a.get("tick")
    if tick is None:
        tick = abs(length) * 0.08 if length else 0.05
    tick_ratio = a.get("tickRatio")
    x, y = float(a["x"]), float(a["y"])
    layers: list[dict[str, Any]] = []

    if along and x2 is not None and y2 is not None:
        x0, y0, x1, y1 = x, y, float(x2), float(y2)
        mid_x = math.exp(0.5 * (math.log(x0) + math.log(x1)))
        mid_y = math.exp(0.5 * (math.log(y0) + math.log(y1)))
        s = float(a.get("capLog") or 0.18)
        layers.append(
            {
                "data": {"values": [{"x": x0, "y": y0, "x2": x1, "y2": y1}]},
                "mark": {"type": "rule", "strokeWidth": sw, "color": color, "strokeCap": "butt"},
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "y": {"field": "y", "type": "quantitative"},
                    "x2": {"field": "x2"},
                    "y2": {"field": "y2"},
                },
            }
        )
        layers.append(
            {
                "data": {
                    "values": [
                        _log_perp_cap(x0, y0, x0, y0, x1, y1, s),
                        _log_perp_cap(x1, y1, x0, y0, x1, y1, s),
                    ]
                },
                "mark": {"type": "rule", "strokeWidth": sw, "color": color},
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "y": {"field": "y", "type": "quantitative"},
                    "x2": {"field": "x2"},
                    "y2": {"field": "y2"},
                },
            }
        )
        if a.get("label"):
            layers.append(
                {
                    "data": {
                        "values": [{"x": mid_x, "y": mid_y * 1.55, "label": a["label"]}]
                    },
                    "mark": {
                        "type": "text",
                        "font": _SERIF,
                        "fontStyle": "normal",
                        "color": color,
                        "align": "center",
                        "baseline": "bottom",
                    },
                    "encoding": {
                        "x": {"field": "x", "type": "quantitative"},
                        "y": {"field": "y", "type": "quantitative"},
                        "text": {"field": "label", "type": "nominal"},
                    },
                }
            )
        return layers

    horizontal = orient == "horizontal"
    if horizontal:
        x0 = x
        x1 = float(a["x2"]) if a.get("x2") is not None else x + length
        mid = (x0 + x1) / 2
        if tick_ratio is not None:
            y_lo, y_hi = y / float(tick_ratio), y * float(tick_ratio)
        else:
            y_lo, y_hi = y - tick, y + tick
        layers.append(
            {
                "data": {"values": [{"x": x0, "x2": x1, "y": y}]},
                "mark": {"type": "rule", "strokeWidth": sw, "color": color, "strokeCap": "butt"},
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "x2": {"field": "x2"},
                    "y": {"field": "y", "type": "quantitative"},
                },
            }
        )
        layers.append(
            {
                "data": {
                    "values": [
                        {"x": x0, "y": y_lo, "y2": y_hi},
                        {"x": x1, "y": y_lo, "y2": y_hi},
                    ]
                },
                "mark": {"type": "rule", "strokeWidth": sw, "color": color},
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "y": {"field": "y", "type": "quantitative"},
                    "y2": {"field": "y2"},
                },
            }
        )
        if a.get("label"):
            layers.append(
                {
                    "data": {"values": [{"x": mid, "y": y, "label": a["label"]}]},
                    "mark": {
                        "type": "text",
                        "dy": -10,
                        "font": _SERIF,
                        "fontStyle": "normal",
                        "color": color,
                        "align": "center",
                        "baseline": "bottom",
                    },
                    "encoding": {
                        "x": {"field": "x", "type": "quantitative"},
                        "y": {"field": "y", "type": "quantitative"},
                        "text": {"field": "label", "type": "nominal"},
                    },
                }
            )
    else:
        y0 = y
        y1 = float(a["y2"]) if a.get("y2") is not None else y + length
        mid = (y0 + y1) / 2
        if tick_ratio is not None:
            x_lo, x_hi = x / float(tick_ratio), x * float(tick_ratio)
        else:
            x_lo, x_hi = x - tick, x + tick
        layers.append(
            {
                "data": {"values": [{"x": x, "y": y0, "y2": y1}]},
                "mark": {"type": "rule", "strokeWidth": sw, "color": color, "strokeCap": "butt"},
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "y": {"field": "y", "type": "quantitative"},
                    "y2": {"field": "y2"},
                },
            }
        )
        layers.append(
            {
                "data": {
                    "values": [
                        {"x": x_lo, "x2": x_hi, "y": y0},
                        {"x": x_lo, "x2": x_hi, "y": y1},
                    ]
                },
                "mark": {"type": "rule", "strokeWidth": sw, "color": color},
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "x2": {"field": "x2"},
                    "y": {"field": "y", "type": "quantitative"},
                },
            }
        )
        if a.get("label"):
            layers.append(
                {
                    "data": {"values": [{"x": x, "y": mid, "label": a["label"]}]},
                    "mark": {
                        "type": "text",
                        "dx": 10,
                        "font": _SERIF,
                        "fontStyle": "normal",
                        "color": color,
                        "align": "left",
                        "baseline": "middle",
                    },
                    "encoding": {
                        "x": {"field": "x", "type": "quantitative"},
                        "y": {"field": "y", "type": "quantitative"},
                        "text": {"field": "label", "type": "nominal"},
                    },
                }
            )
    return layers


def _arrow_layers(a: Mapping[str, Any]) -> list[dict[str, Any]]:
    color = a.get("color") or "#14271d"
    sw = a.get("strokeWidth") or 1.6
    tip_size = a.get("tipSize") or 55
    x, y = float(a["x"]), float(a["y"])
    x2, y2 = float(a["x2"]), float(a["y2"])
    angle = math.degrees(math.atan2(y2 - y, x2 - x)) + 90
    layers: list[dict[str, Any]] = [
        {
            "data": {"values": [{"x": x, "y": y, "x2": x2, "y2": y2}]},
            "mark": {"type": "rule", "strokeWidth": sw, "color": color, "strokeCap": "round"},
            "encoding": {
                "x": {"field": "x", "type": "quantitative"},
                "y": {"field": "y", "type": "quantitative"},
                "x2": {"field": "x2"},
                "y2": {"field": "y2"},
            },
        },
        {
            "data": {"values": [{"x": x2, "y": y2, "angle": angle}]},
            "mark": {
                "type": "point",
                "shape": "triangle",
                "filled": True,
                "size": tip_size,
                "color": color,
            },
            "encoding": {
                "x": {"field": "x", "type": "quantitative"},
                "y": {"field": "y", "type": "quantitative"},
                "angle": {"field": "angle", "type": "quantitative"},
            },
        },
    ]
    if a.get("label"):
        layers.append(
            {
                "data": {
                    "values": [
                        {"x": (x + x2) / 2, "y": (y + y2) / 2, "label": a["label"]}
                    ]
                },
                "mark": {
                    "type": "text",
                    "dy": -8,
                    "fontSize": 11,
                    "color": color,
                    "align": "center",
                    "baseline": "bottom",
                },
                "encoding": {
                    "x": {"field": "x", "type": "quantitative"},
                    "y": {"field": "y", "type": "quantitative"},
                    "text": {"field": "label", "type": "nominal"},
                },
            }
        )
    return layers


def annotation_layers(annotations: Sequence[Mapping[str, Any]]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for a in annotations:
        kind = a.get("kind")
        if kind == "scaleBar":
            out.extend(_scale_bar_layers(a))
        elif kind == "arrow":
            out.extend(_arrow_layers(a))
    return out


def with_annotations(
    spec: dict[str, Any],
    annotations: Sequence[Mapping[str, Any]] | None,
) -> dict[str, Any]:
    """Merge annotation layers into a unit or layered Vega-Lite spec."""
    if not annotations:
        return spec
    extra = annotation_layers(annotations)
    if not extra:
        return spec
    if isinstance(spec.get("layer"), list):
        return {**spec, "layer": [*spec["layer"], *extra]}
    if "mark" in spec or "encoding" in spec:
        base = {
            k: spec[k]
            for k in ("data", "mark", "encoding", "transform", "params")
            if k in spec
        }
        rest = {k: v for k, v in spec.items() if k not in base}
        return {**rest, "layer": [base, *extra]}
    return {**spec, "layer": extra}


def take_annotations(spec: dict[str, Any]) -> tuple[dict[str, Any], list[Annotation]]:
    """Pull top-level ``annotations`` off a payload; return ``(spec, list)``."""
    raw = spec.get("annotations")
    if not isinstance(raw, list) or not raw:
        cleaned = {k: v for k, v in spec.items() if k != "annotations"}
        return cleaned, []
    annotations: list[Annotation] = [
        a  # type: ignore[misc]
        for a in raw
        if isinstance(a, dict) and a.get("kind") in ("scaleBar", "arrow")
    ]
    cleaned = {k: v for k, v in spec.items() if k != "annotations"}
    return cleaned, annotations

File: core/src/annotations.ts
import type { VegaLiteSpec } from "./specs";

/**
 * Chart annotations — portable Vega-Lite layers for paper/web.
 *
 * - {@link ScaleBarAnnotation}: a `|———|` scale bar in data coordinates
 * - {@link ArrowAnnotation}: a directed arrow (rule + tip) with optional label
 *
 * Pass via `withAnnotations(spec, …)` or a top-level `annotations` array on a
 * RawChart / fence spec (stripped before vega-embed).
 */

export type ScaleBarAnnotation = {
  kind: "scaleBar";
  /** Start of the bar (data coords). */
  x: number;
  y: number;
  /**
   * Length along a horizontal/vertical axis. Ignored when both
   * {@link x2} and {@link y2} are set (matplotlib-style chord on a curve).
   */
  length?: number;
  label?: string;
  color?: string;
  strokeWidth?: number;
  /**
   * `horizontal` / `vertical` axis-aligned bars, or `along` for a free
   * segment (x,y)→(x2,y2) with end-caps ⊥ in log–log space (default when
   * both ends are given).
   */
  orientation?: "horizontal" | "vertical" | "along";
  /** Linear end-cap half-length (axis-aligned modes). */
  tick?: number;
  /** Geometric end-cap half-span on log axes (axis-aligned modes). */
  tickRatio?: number;
  /** End x (with y2: chord along curve). */
  x2?: number;
  /** End y (with x2: chord along curve). */
  y2?: number;
  /**
   * Log-space half-length of end-caps for `along` chords (default 0.18).
   * Caps are perpendicular to the segment in (ln x, ln y).
   */
  capLog?: number;
};

export type ArrowAnnotation = {
  kind: "arrow";
  /** Tail (start). */
  x: number;
  y: number;
  /** Tip (end). */
  x2: number;
  y2: number;
  label?: string;
  color?: string;
  strokeWidth?: number;
  /** Tip mark size (Vega point size, area in px²). Default 55. */
  tipSize?: number;
};

export type Annotation = ScaleBarAnnotation | ArrowAnnotation;

function cleanLayer(layer: Record<string, unknown>): VegaLiteSpec {
  for (const k of Object.keys(layer)) {
    if (layer[k] === undefined) delete layer[k];
  }
  return layer as VegaLiteSpec;
}

/** End-cap endpoints ⊥ to (x0,y0)→(x1,y1) in log–log space (matplotlib-like). */
function logPerpCap(
  x: number,
  y: number,
  x0: number,
  y0: number,
  x1: number,
  y1: number,
  s: number,
): { x: number; y: number; x2: number; y2: number } {
  const dx = Math.log(x1 / x0);
  const dy = Math.log(y1 / y0);
  const n = Math.hypot(dx, dy) || 1;
  const px = (-dy / n) * s;
  const py = (dx / n) * s;
  return {
    x: Math.exp(Math.log(x) + px),
    y: Math.exp(Math.log(y) + py),
    x2: Math.exp(Math.log(x) - px),
    y2: Math.exp(Math.log(y) - py),
  };
}

function scaleBarLayers(a: ScaleBarAnnotation): VegaLiteSpec[] {
  const color = a.color ?? "#14271d";
  const sw = a.strokeWidth ?? 1.8;
  const font = "Times New Roman, Times, STIX Two Text, STIXGeneral, serif";
  const layers: VegaLiteSpec[] = [];

  // Matplotlib-style chord on a curve: both ends fully specified.
  const along =
    a.orientation === "along" ||
    (a.x2 != null &&
      a.y2 != null &&
      a.orientation !== "horizontal" &&
      a.orientation !== "vertical");

  if (along && a.x2 != null && a.y2 != null) {
    const x0 = a.x;
    const y0 = a.y;
    const x1 = a.x2;
    const y1 = a.y2;
    const midX = Math.exp(0.5 * (Math.log(x0) + Math.log(x1)));
    const midY = Math.exp(0.5 * (Math.log(y0) + Math.log(y1)));
    const s = a.capLog ?? 0.18;
    const c0 = logPerpCap(x0, y0, x0, y0, x1, y1, s);
    const c1 = logPerpCap(x1, y1, x0, y0, x1, y1, s);
    layers.push(
      cleanLayer({
        data: { values: [{ x: x0, y: y0, x2: x1, y2: y1 }] },
        mark: { type: "rule", strokeWidth: sw, color, strokeCap: "butt" },
        encoding: {
          x: { field: "x", type: "quantitative" },
          y: { field: "y", type: "quantitative" },
          x2: { field: "x2" },
          y2: { field: "y2" },
        },
      }),
      cleanLayer({
        data: { values: [c0, c1] },
        mark: { type: "rule", strokeWidth: sw, color },
        encoding: {
          x: { field: "x", type: "quantitative" },
          y: { field: "y", type: "quantitative" },
          x2: { field: "x2" },
          y2: { field: "y2" },
        },
      }),
    );
    if (a.label) {
      layers.push(
        cleanLayer({
          data: {
            values: [{ x: midX, y: midY * 1.55, label: a.label }],
          },
          mark: {
            type: "text",
            font,
            fontStyle: "normal",
            color,
            align: "center",
            baseline: "bottom",
          },
          encoding: {
            x: { field: "x", type: "quantitative" },
            y: { field: "y", type: "quantitative" },
            text: { field: "label", type: "nominal" },
          },
        }),
      );
    }
    return layers;
  }

  const horizontal = (a.orientation ?? "horizontal") === "horizontal";

  if (horizontal) {
    const x0 = a.x;
    const x1 =
      a.x2 ?? a.x + (Number.isFinite(a.length) ? (a.length as number) : 0);
    const mid = (x0 + x1) / 2;
    const yLo =
      a.tickRatio != null
        ? a.y / a.tickRatio
        : a.y -
          (a.tick ??
            (Number.isFinite(a.length) && a.length
              ? Math.abs(a.length as number) * 0.08
              : 0.05));
    const yHi =
      a.tickRatio != null
        ? a.y * a.tickRatio
        : a.y +
          (a.tick ??
            (Number.isFinite(a.length) && a.length
              ? Math.abs(a.length as number) * 0.08
              : 0.05));
    layers.push(
      cleanLayer({
        data: { values: [{ x: x0, x2: x1, y: a.y }] },
        mark: { type: "rule", strokeWidth: sw, color, strokeCap: "butt" },
        encoding: {
          x: { field: "x", type: "quantitative" },
          x2: { field: "x2" },
          y: { field: "y", type: "quantitative" },
        },
      }),
      cleanLayer({
        data: {
          values: [
            { x: x0, y: yLo, y2: yHi },
            { x: x1, y: yLo, y2: yHi },
          ],
        },
        mark: { type: "rule", strokeWidth: sw, color },
        encoding: {
          x: { field: "x", type: "quantitative" },
          y: { field: "y", type: "quantitative" },
          y2: { field: "y2" },
        },
      }),
    );
    if (a.label) {
      layers.push(
        cleanLayer({
          data: { values: [{ x: mid, y: a.y, label: a.label }] },
          mark: {
            type: "text",
            dy: -10,
            font,
            fontStyle: "normal",
            color,
            align: "center",
            baseline: "bottom",
          },
          encoding: {
            x: { field: "x", type: "quantitative" },
            y: { field: "y", type: "quantitative" },
            text: { field: "label", type: "nominal" },
          },
        }),
      );
    }
  } else {
    const y0 = a.y;
    const y1 =
      a.y2 ?? a.y + (Number.isFinite(a.length) ? (a.length as number) : 0);
    const mid = (y0 + y1) / 2;
    const tick =
      a.tick ??
      (Number.isFinite(a.length) && a.length
        ? Math.abs(a.length as number) * 0.08
        : 0.05);
    const xLo = a.tickRatio != null ? a.x / a.tickRatio : a.x - tick;
    const xHi = a.tickRatio != null ? a.x * a.tickRatio : a.x + tick;
    layers.push(
      cleanLayer({
        data: { values: [{ x: a.x, y: y0, y2: y1 }] },
        mark: { type: "rule", strokeWidth: sw, color, strokeCap: "butt" },
        encoding: {
          x: { field: "x", type: "quantitative" },
          y: { field: "y", type: "quantitative" },
          y2: { field: "y2" },
        },
      }),
      cleanLayer({
        data: {
          values: [
            { x: xLo, x2: xHi, y: y0 },
            { x: xLo, x2: xHi, y: y1 },
          ],
        },
        mark: { type: "rule", strokeWidth: sw, color },
        encoding: {
          x: { field: "x", type: "quantitative" },
          x2: { field: "x2" },
          y: { field: "y", type: "quantitative" },
        },
      }),
    );
    if (a.label) {
      layers.push(
        cleanLayer({
          data: { values: [{ x: a.x, y: mid, label: a.label }] },
          mark: {
            type: "text",
            dx: 10,
            font,
            fontStyle: "normal",
            color,
            align: "left",
            baseline: "middle",
          },
          encoding: {
            x: { field: "x", type: "quantitative" },
            y: { field: "y", type: "quantitative" },
            text: { field: "label", type: "nominal" },
          },
        }),
      );
    }
  }

  return layers;
}

function arrowLayers(a: ArrowAnnotation): VegaLiteSpec[] {
  const color = a.color ?? "#14271d";
  const sw = a.strokeWidth ?? 1.6;
  const tipSize = a.tipSize ?? 55;
  // Vega point `angle` is degrees clockwise from north; atan2 is from +x CCW.
  const angleDeg = (Math.atan2(a.y2 - a.y, a.x2 - a.x) * 180) / Math.PI + 90;

  const layers: VegaLiteSpec[] = [
    cleanLayer({
      data: {
        values: [{ x: a.x, y: a.y, x2: a.x2, y2: a.y2 }],
      },
      mark: { type: "rule", strokeWidth: sw, color, strokeCap: "round" },
      encoding: {
        x: { field: "x", type: "quantitative" },
        y: { field: "y", type: "quantitative" },
        x2: { field: "x2" },
        y2: { field: "y2" },
      },
    }),
    cleanLayer({
      data: {
        values: [{ x: a.x2, y: a.y2, angle: angleDeg }],
      },
      mark: {
        type: "point",
        shape: "triangle",
        filled: true,
        size: tipSize,
        color,
      },
      encoding: {
        x: { field: "x", type: "quantitative" },
        y: { field: "y", type: "quantitative" },
        angle: { field: "angle", type: "quantitative" },
      },
    }),
  ];

  if (a.label) {
    const mx = (a.x + a.x2) / 2;
    const my = (a.y + a.y2) / 2;
    layers.push(
      cleanLayer({
        data: { values: [{ x: mx, y: my, label: a.label }] },
        mark: {
          type: "text",
          dy: -8,
          fontSize: 11,
          color,
          align: "center",
          baseline: "bottom",
        },
        encoding: {
          x: { field: "x", type: "quantitative" },
          y: { field: "y", type: "quantitative" },
          text: { field: "label", type: "nominal" },
        },
      }),
    );
  }

  return layers;
}

/** Expand annotations into Vega-Lite unit layers (own data, no shared scales). */
export function annotationLayers(annotations: Annotation[]): VegaLiteSpec[] {
  const out: VegaLiteSpec[] = [];
  for (const a of annotations) {
    if (a.kind === "scaleBar") out.push(...scaleBarLayers(a));
    else if (a.kind === "arrow") out.push(...arrowLayers(a));
  }
  return out;
}

/**
 * Merge annotation layers into a unit or layered Vega-Lite spec.
 * Unit specs become layered; existing `layer` arrays are appended to.
 */
export function withAnnotations(
  spec: VegaLiteSpec,
  annotations: Annotation[] | undefined | null,
): VegaLiteSpec {
  if (!annotations?.length) return spec;
  const extra = annotationLayers(annotations);
  if (!extra.length) return spec;

  if (Array.isArray(spec.layer)) {
    return {
      ...spec,
      layer: [...(spec.layer as VegaLiteSpec[]), ...extra],
    };
  }

  // Unit → layered. Keep top-level data only if the unit layer still needs it
  // via inheritance; copy unit fields into layer[0] for a self-contained base.
  const { mark, encoding, data, transform, params, ...rest } = spec as Record<
    string,
    unknown
  >;

  if (mark !== undefined || encoding !== undefined) {
    const baseLayer: VegaLiteSpec = {};
    if (data !== undefined) baseLayer.data = data;
    if (mark !== undefined) baseLayer.mark = mark;
    if (encoding !== undefined) baseLayer.encoding = encoding;
    if (transform !== undefined) baseLayer.transform = transform;
    if (params !== undefined) baseLayer.params = params;
    return {
      ...rest,
      layer: [baseLayer, ...extra],
    } as VegaLiteSpec;
  }

  // Already a multi-view / facet without unit marks — attach layers only.
  return {
    ...spec,
    layer: extra,
  };
}

/**
 * Pull a top-level `annotations` key off a fence/RawChart payload and merge.
 * Returns `{ spec, annotations }` where `spec` has no `annotations` field.
 */
export function takeAnnotations(spec: VegaLiteSpec): {
  spec: VegaLiteSpec;
  annotations: Annotation[];
} {
  const raw = (spec as { annotations?: unknown }).annotations;
  if (!Array.isArray(raw) || raw.length === 0) {
    if ("annotations" in spec) {
      const { annotations: _drop, ...rest } = spec as Record<string, unknown>;
      return { spec: rest as VegaLiteSpec, annotations: [] };
    }
    return { spec, annotations: [] };
  }
  const annotations = raw.filter(
    (a): a is Annotation =>
      !!a &&
      typeof a === "object" &&
      ((a as Annotation).kind === "scaleBar" ||
        (a as Annotation).kind === "arrow"),
  );
  const { annotations: _drop, ...rest } = spec as Record<string, unknown>;
  return { spec: rest as VegaLiteSpec, annotations };
}
